tests/pyspark_delta/test_pyspark_writer.py

```diff
@@ -0,0 +1,54 @@
+import pytest
+import chispa
+import pyspark
+from delta import *
+import shutil
+import os
+
+builder = (
+    pyspark.sql.SparkSession.builder.appName("MyApp")
+    .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension")
+    .config(
+        "spark.sql.catalog.spark_catalog",
+        "org.apache.spark.sql.delta.catalog.DeltaCatalog",
+    )
+)
+
+spark = configure_spark_with_delta_pip(builder).getOrCreate()
+
+def test_reference_table_1_append():
+    os.makedirs("tmp")
+
+    # copy over existing reference table
+    shutil.copytree("out/tables/generated/reference_table_1/delta", "tmp/delta")
+    df = spark.read.format("delta").load("tmp/delta")
+    # +------+------+-------+                                                         
+    # |letter|number|a_float|
+    # +------+------+-------+
+    # |     a|     1|    1.1|
+    # |     b|     2|    2.2|
+    # |     c|     3|    3.3|
+    # |     d|     4|    4.4|
+    # |     e|     5|    5.5|
+    # +------+------+-------+
+    
+    # append data to the reference table
+    rdd = spark.sparkContext.parallelize([("z", 9, 9.9)])
+    df = rdd.toDF(["letter", "number", "a_float"])
+    df.write.format("delta").mode("append").save("tmp/delta")
+
+    # created expected content
+    rdd = spark.sparkContext.parallelize([
+        ("a", 1, 1.1),
+        ("b", 2, 2.2),
+        ("c", 3, 3.3),
+        ("d", 4, 4.4),
+        ("e", 5, 5.5),
+        ("z", 9, 9.9),
+    ])
+    expected_df = rdd.toDF(["letter", "number", "a_float"])    
+
+    # make assertions & cleanup
+    actual_df = spark.read.format("delta").load("tmp/delta")
+    chispa.assert_df_equality(actual_df, expected_df, ignore_row_order=True)
```

Comment:
Besides asserting the result is right, we also need to validate the log. First, there are probably general invariants we want to check. Ones that aren't specific to this operation, but should always be true after every write. A few examples:

1. Only one new log entry was added.
2. All log entries prior are still there and unchanged.
3. Old data files referenced by the log are still there and unchanged.
4. The new log entry is valid according to the protocol. (Perhaps we can generate a JSON schema for the delta log entries and use that as a first pass?)

Beyond that, we might also want operation-specific validation. Like an "append" operation has `dataChange=True` on all `add` actions and doesn't contain any `remove` actions?

Reply:
Yea, let's brainstorm how to write this code.  Parsing the transaction log from scratch in Python would seem a little tedious.  Perhaps I can used delta-rs for this?

Reply:
Wouldn't using delta-rs in this way be circular? We could just compare the JSON files directly... some more thought needed for the checkpoint files.

Reply:
@nkarpov - yea, using delta-rs in that way would definitely be circular, especially for testing the delta-rs connector.  Less so for testing the Flink, Hive, etc. connectors.

So I guess we just make a Python reader for parsing all the intricacies of the transaction log.  That seems kind of complicated as well.

I really don't know the best path forward here.